src/browser/subscription_manager.ts

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2017 OpenFin Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import ofEvents from './of_events';
+import route from '../common/route';
+import { Identity } from '../shapes';
+
+interface Subscription {
+    fn: () => void;
+    refCount: number;
+}
+
+export default class SubscriptionManager {
+    private subscriptionList: Map<string, Map<string, Subscription>>;
+
+    constructor() {
+        this.subscriptionList = new Map();
+
+        ofEvents.on(route.window('closed'), (identity: Identity) => {
+            this.removeAllSubscriptions(identity);
+        });
+
+        ofEvents.on(route('externalconn', 'closed'), (identity: Identity) => {
+            this.removeAllSubscriptions(identity);
+        });
+    }
+
+    private identityToKey(identity: Identity): string {
+        return encodeURIComponent(`${identity.uuid}/${identity.name}`);
+    }
+
+    private genSubscriptionKey(...args: any[]): string {
+        const stringArgs = args.reduce((prev, curr) => `${curr}${prev}`);
+        return encodeURIComponent(stringArgs);
+    }
+
+    public subscriptionExists(identity: Identity, ...args: any[]): boolean {
+        const key = this.genSubscriptionKey.apply(null, args);
+        const identityKey = this.identityToKey(identity);
+        const identitySubs = this.subscriptionList.get(identityKey);
+
+        return !!identitySubs && identitySubs.has(key);
+    }
+
+    public uppSubscriptionRefCount(identity: Identity, ...args: any[]): void {
+        const key = this.genSubscriptionKey.apply(null, args);
+        const identityKey = this.identityToKey(identity);
+
+        this.subscriptionList.get(identityKey).get(key).refCount++;
+    }
+
+    public registerSubscription(fn: () => void, identity: Identity, ...args: any[]): void {
+        const key = this.genSubscriptionKey.apply(null, args);
```

Comment:
Lets declare all these public functions with  arrow key to prevent loss of context.